### DISCUSSION TITLE
Add an access to sipsession from external lib to allow i.e. sipsess_reject

### DISCUSSION THREAD

**Discussion by zucher** [baresip/baresip#3207] [Ideas]
Hello,

as I want to setup in ringing diversion ( quite similar to #616 ) into my application. I had to create an accessor  to help to retrieve the session associated to a call ( struct call model not accessible from  external code )

is it possible to add the following declaration into the baresip call part ?

```c
struct sipsess * get_sipsess(const struct call* call )
{
	return call ? call->sess : NULL;
}
```

for the following finl usage:
```c
	(void)sipsess_reject(get_sipsess(call), 302, "Moved Temporarily",
			     "Contact: <%s>%s\r\n"
			     "Diversion: <%s>%s\r\n"
			     "Content-Length: 0\r\n\r\n"
			     ,
			     contact, contact_estr ? contact_estr : "", account_aor(ua_account(call_get_ua(call))), estr ? estr : "");
```

where `contact` is the target divertion URI and diversion the current UA reference.
Some SIP phone are using this reject method to divert incoming call ( such as microsip )
Such cases described in the RFC: [https://www.rfc-editor.org/rfc/rfc5806](https://www.rfc-editor.org/rfc/rfc5806)

**Comment by cspiel1** [comment]
Currently there is no real need to directly call `sipsess_reject()` from application level. There are two different cases: You want to cancel the incoming call
- (A) when there is already a call object. Then `call_hangup()` should be used in order to reject the call.
- (B) You want to cancel before there is a call object and a SIP session created. Then your application module, or your application that uses libbaresip should register a bevent handler and handle the `case UA_EVENT_SIPSESS_CONN`. See module/menu.c and commit 8e73f5beebef8dbfffe0f06c0fde1ff1509866b0 how this can be done! Short: Use `sip_treply()` in this case!

In case of (A), note that the call object also have to do its cleanup, and the `sipsess` itself needs a `mem_deref`. The function `call_hangup()` does this correctly.

**Comment by zucher** [comment]
@cspiel1 ,
Thank you for pointing me the reference, I have to investigate your proposition around the UA_EVENT_SIPSESS_CONN handling.

**Comment by zucher** [comment]
@cspiel1 , 
After review, for the case A, the call_hangup method doesn't provide enough capabilities to enhance the response with additional headers, such as the "Contact" and "Diversion".
for the case B, as part of the application is running in an external process with the bus communication, managing interaction on  UA_EVENT_SIPSESS_CONN becomes difficult. The destination of the diversion is dynamical and have to be resolved only at incoming call reception.

**Comment by cspiel1** [comment]
Some earlier work on this feature was done and rejected:
- https://github.com/baresip/baresip/pull/2595
- https://github.com/baresip/baresip/pull/2596

The feature would also be useful for us in order to have a tool for testing SIP 302. I'll try to find a solution with minimal change in the baresip core.

**Comment by cspiel1** [comment]
Started to work on this feature request:
https://github.com/baresip/baresip-apps/pull/57

Not tested so far.

**Comment by cspiel1** [comment]
@zucher now https://github.com/baresip/baresip-apps/pull/57 is ready for review and testing. Please let me know how it fits for your needs!

**Comment by zucher** [comment]
Tks, I try to check tomorrow
